collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListViewModel.kt

```diff
@@ -187,6 +186,9 @@ class BlankFormListViewModel(
             syncRepository.isSyncing(projectId)
         ) { isSyncing ->
             isSyncingWithServerRunning.value = isSyncing
+            if (!isSyncing) {
+                loadFromDatabase()
```

Comment:
We really shouldn't have code with side effects in `map` calls in general (as it can cause unexpected results), but it's definitely a problem here due to `isSyncingWithServer` returning a new `LiveData` as the result of a `map` call every time. This would mean that when the value of `syncRepository.isSyncing(projectId)` switches from `true` to `false` we would call the side effects (`loadFormDatabase` and the update to `isSyncingWithServerRunning`) once for every time we've called `isSyncingWithServer`. I'd completely missed the interaction with  `isSyncingWithServerRunning` in here when we introduced the `BlankFormListViewModel`, but should have brought this up then.

This implementation initially confused me as well as it requires `isSyncingWithServer` to have been called (and the result observed) at least once for `formsToDisplay` or `isLoading` to be updated correctly which feels surprising. Here's a test that demonstrates the problem:

```kotlin
@Test
fun `isLoading is true while syncing`() {
    val syncing = MutableLiveData(false)
    whenever(syncRepository.isSyncing(projectId)).thenReturn(syncing)

    createViewModel()
    assertThat(viewModel.isLoading.getOrAwaitValue(), equalTo(false))

    syncing.value = true
    // viewModel.isSyncingWithServer().observeForever { }
    assertThat(viewModel.isLoading.getOrAwaitValue(), equalTo(true))
}
```

That test fails, but it will pass if I uncomment the second last line. This will all work in the app as we happen to call the right methods in the right order, but it could definitely cause problems later on if we move things around. I think should move things around so that `isLoading` and `formsToDisplay` aren't dependent on `isSyncingWithServer` being called. A few thoughts on that:

- Why do we need `isSyncingWithServer` and `isLoading`? I feel like there's an answer on the tip of my tongue, but I can't quite get to it.
- We can probably just use `SyncRepository#isSyncing` in the zip that defines `isLoading` instead of needing a separate `isSyncingWithServerRunning` field.
- To fix #5331, I think `SyncRepository#isSyncing` could be used with `observerForever`/`removeObserver`/`ViewModel#onCleared` to perform a db load whenever there isn't a sync in progress from `BlankFormListViewModel`. We definitely want a structure like `InstancesAppState` in place for blank forms down the line. Not sure if that's the easiest thing right now though.
- As far as I can see there aren't any tests for `isLoading` at the moment. We should add those (like the above example) here.

Reply:
> Why do we need isSyncingWithServer and isLoading? I feel like there's an answer on the tip of my tongue, but I can't quite get to it.

We need both `isSyncingWithServer` and `isOutOfSyncWithServer` because based on changes we update icons in `BlankFormListMenuDelegate`. 

> We can probably just use SyncRepository#isSyncing in the zip that defines isLoading instead of needing a separate isSyncingWithServerRunning field.

No because as I said above it's used in `BlankFormListMenuDelegate` as well.

> We really shouldn't have code with side effects in map calls in general (as it can cause unexpected results)

I thought the same but as this structure already exist and is needed (see above) I decided to use it.